pkg/apis/admissionregistration/validation/validation.go

```diff
@@ -760,6 +792,36 @@ func validateValidation(v *admissionregistration.Validation, paramKind *admissio
 	return allErrors
 }
 
+func validateAuditAnnotation(meta metav1.ObjectMeta, v *admissionregistration.AuditAnnotation, paramKind *admissionregistration.ParamKind, fldPath *field.Path) field.ErrorList {
+	var allErrors field.ErrorList
+	if len(meta.GetName()) != 0 {
+		name := meta.GetName()
+		allErrors = append(allErrors, apivalidation.ValidateQualifiedName(name+"/"+v.Key, fldPath.Child("key"))...)
+	} else {
+		allErrors = append(allErrors, field.Invalid(fldPath.Child("key"), v.Key, "requires metadata.name be non-empty"))
+	}
+
+	trimmedValueExpression := strings.TrimSpace(v.ValueExpression)
+	if len(trimmedValueExpression) == 0 {
+		allErrors = append(allErrors, field.Required(fldPath.Child("valueExpression"), "valueExpression is not specified"))
+	} else {
+		result := plugincel.CompileValidatingPolicyAuditAnnotationValueExpression(trimmedValueExpression, paramKind != nil)
```

Comment:
can we ensure the expression returns string or null or dynamic-type?

Reply:
I double checked and we're requiring the expression evaluates to a known type. I had forgotten we are strict about this.

If a dynamic value is returned, it can be cast to appease the type check. All scalars cast to string, so `string(whatever)` works. Also, when we add `string.format()` it will be even easier because `"%d".format([dynamicIntVal])` will be handled right at runtime.

Reply:
This is properly documented.